### PR TITLE
Replace `pdf_path` & `gs_stack` with `Vec`

### PR DESCRIPTION
**Pull request by burrbull**
use `Option<CString>` for `spot_color_name` in `pdf_color` and other changes in `dpx_pdfdraw.rs`

~Note: `issue393_ungetc` test doesn't pass now. Panics on empty `gs_stack`. I think it is not my fault.~

### CODE REVIEW

**Comment by burrbull** [comment]
Fixed

File: engine/src/dpx_pdfdev.rs [hunk 1]
```diff
@@ -236,6 +236,16 @@ impl pdf_tmatrix {
             f: 0.,
         }
     }
+    pub const fn init() -> Self {
```
Comment (crlf0710) [inline engine/src/dpx_pdfdev.rs]: `identity` or something like that

Comment (burrbull) [reply engine/src/dpx_pdfdev.rs]: Done.

**Comment by crlf0710** [comment]
Do you mind leaving this PR pending for a few days... There're some structural changes that i want to make first. And the state management is too much relying on the global variables (Thus the `Mutex`). I think we'll want to change that.

File: dpx/src/dpx_pdfdev.rs [hunk 2]
```diff
@@ -266,6 +276,12 @@ impl pdf_coord {
         Self { x: 0., y: 0. }
     }
 }
+impl PartialEq for pdf_coord {
```
Comment (crlf0710) [inline dpx/src/dpx_pdfdev.rs]: I think this PartialEq is very weird... Do you think make it a method is a better idea?

File: dpx/src/dpx_pdfdraw.rs [hunk 3]
```diff
@@ -965,239 +793,176 @@ unsafe extern "C" fn pdf_dev__flushpath(
...
+
+impl pdf_gstate {
+    pub fn init() -> Self {
+        Self {
+            cp: pdf_coord::new(),
+            matrix: pdf_tmatrix::identity(),
+            strokecolor: pdf_color_graycolor_new(0.).unwrap(),
+            fillcolor: pdf_color_graycolor_new(0.).unwrap(),
+            linedash: LineDash::default(),
+            linecap: 0,
```
Comment (crlf0710) [inline dpx/src/dpx_pdfdraw.rs]: line cap and line join can be turned into enum in a future pr.

File: dpx/src/dpx_pdfdraw.rs [hunk 4]
```diff
@@ -965,239 +793,176 @@ unsafe extern "C" fn pdf_dev__flushpath(
...
-    assert!(!stack.is_null());
-    if (*stack).size == 0i32 {
-        return 0 as *mut libc::c_void;
+}
+
+use lazy_static::lazy_static;
+use std::sync::Mutex;
+
+lazy_static! {
+    static ref gs_stack: Mutex<Vec<pdf_gstate>> = Mutex::new({
```
Comment (crlf0710) [inline dpx/src/dpx_pdfdraw.rs]: i'm not very happy with the Mutex here. Can we add a struct called PdfDev or pdf_dev and collect these global states within it. and make the drawing invocations its method? (might be a big change to the code base)

File: dpx/src/dpx_pdfdev.rs (hunk 2 above)
Comment (cormacrelf) [reply dpx/src/dpx_pdfdev.rs]: More precisely, it’s symmetric but not transitive, so you shouldn’t implement PartialEq this way.

File: dpx/src/dpx_pdfdraw.rs (hunk 4 above)
Comment (cormacrelf) [reply dpx/src/dpx_pdfdraw.rs]: Helpful for seeing which statics are linked to by C++:

```sh
rg 'static mut (\w+):' myfile -o -r '$1' | xargs -n1 -I {} sh -c 'echo {} && rg -F {} c_sources_directory'
```

Generally this helps for moving definitions out of extern “C” when they don’t need to be in one.

Comment (burrbull) [reply dpx/src/dpx_pdfdraw.rs]: > i'm not very happy with the Mutex here. Can we add a struct called PdfDev or pdf_dev and collect these global states within it. and make the drawing invocations its method? (might be a big change to the code base)

I thought about this variant too. The problem as you say, it is a big change. And most likely we will be able to do this at one of the last steps of "oxidation".

File: dpx/src/dpx_pdfdev.rs (hunk 2 above)
Comment (burrbull) [reply dpx/src/dpx_pdfdev.rs]: See last commit.

File: dpx/src/dpx_pdfdraw.rs (hunk 4 above)
Comment (burrbull) [reply dpx/src/dpx_pdfdraw.rs]: Mutex moved to start of file. Comment added.

**Comment by crlf0710** [comment]
Let's merge it, i think.

File: dpx/src/dpx_pdfdev.rs (hunk 2 above)
Comment (crlf0710) [reply dpx/src/dpx_pdfdev.rs]: PartialEq actually doesn't have to be transitive, Eq have to be. But it's still better to not hide these kind of logic under equal operator。

File: dpx/src/dpx_pdfdraw.rs (hunk 4 above)
Comment (burrbull) [reply dpx/src/dpx_pdfdraw.rs]: https://github.com/rust-lang/rust/pull/64028

**Pull request merged**